# 3 Business Owner Perspective

In this notebook, we will analyze the taxi data from the perspective of a business owner who is interested in establishing a retail outlet in New York City. The business owner wants to determine the optimal store location, ideally on a busy street that receives a lot of taxi traffic.

We will start with a bird's eye view of Manhattan, and examine general taxi traffic on a map that ranks the various taxi zones based on dropoffs. We will use this view to pick a general neighborhood.

Once we pick the general zone we want to be in, we will get down to the street level. We will construct a detailed street map with individual dropoffs plotted on it. 

Then, we will use machine learning to discover hotspots of activity. This will be achieved by applying a clustering algorithm (unsupervised machine learning) to Latitude and Longitude information contained in the data. The resulting clusters will become the short list of potential locations for the business owner's new store!

In [13]:
#The bird's eye view is constructed by using summarizing the data by dropoff zone, then visualizing it using QGIS
#We have already summarized the data by dropoff zone in Step 1, so we will pull in that information here.
import pandas as pd
df = pd.read_csv('../01_Tourist_Resident/01_Tourist_Resident.csv')
df

ROWID               dropoff_zone do_weekday  do_hour  ridecount
0          0             Midtown Center  Wednesday      7.0      10075
1          1             Midtown Center   Thursday      7.0       9825
2          2             Midtown Center  Wednesday      8.0       9466
3          3             Midtown Center   Thursday      8.0       9289
4          4             Midtown Center  Wednesday      9.0       7932
5          5             Midtown Center    Tuesday      7.0       7920
6          6               Midtown East  Wednesday      8.0       7603
7          7             Midtown Center   Thursday      9.0       7515
8          8               Midtown East  Wednesday      7.0       7372
9          9             Midtown Center    Tuesday      8.0       7367
10        10               Midtown East   Thursday      8.0       7291
11        11             Midtown Center     Monday      7.0       7282
12        12             Midtown Center     Friday      8.0       7244
13        13             Midtown Center     Friday      7.0       7219
14        14               Midtown East   Thursday      7.0       7163
15        15             Midtown Center     Monday      8.0       6763
16        16  Times Sq/Theatre District     Friday     19.0       6142
17        17             Midtown Center    Tuesday      9.0       6130
18        18             Midtown Center  Wednesday     10.0       6098
19        19             Midtown Center     Friday      9.0       6018
20        20               Midtown East    Tuesday      8.0       5825
21        21               Midtown East    Tuesday      7.0       5820
22        22               Midtown East  Wednesday      9.0       5625
23        23             Midtown Center   Thursday     10.0       5612
24        24             Midtown Center  Wednesday     12.0       5496
25        25               Midtown East   Thursday      9.0       5491
26        26             Midtown Center     Monday      9.0       5476
27        27             Midtown Center  Wednesday     11.0       5459
28        28               Midtown East     Monday      7.0       5407
29        29               Midtown East     Friday      8.0       5385
...      ...                        ...        ...      ...        ...
39842  39842              Willets Point   Thursday      9.0          1
39843  39843              Willets Point   Thursday     12.0          1
39844  39844              Willets Point   Thursday     16.0          1
39845  39845              Willets Point   Thursday     17.0          1
39846  39846              Willets Point   Thursday     18.0          1
39847  39847              Willets Point   Thursday     19.0          1
39848  39848              Willets Point     Friday     12.0          1
39849  39849              Willets Point     Friday     14.0          1
39850  39850              Willets Point     Friday     19.0          1
39851  39851              Willets Point     Friday     22.0          1
39852  39852              Willets Point   Saturday      2.0          1
39853  39853              Willets Point   Saturday     22.0          1
39854  39854   Williamsbridge/Olinville     Sunday      8.0          1
39855  39855   Williamsbridge/Olinville     Monday      4.0          1
39856  39856   Williamsbridge/Olinville    Tuesday      4.0          1
39857  39857   Williamsbridge/Olinville    Tuesday      8.0          1
39858  39858   Williamsbridge/Olinville    Tuesday     15.0          1
39859  39859   Williamsbridge/Olinville  Wednesday      6.0          1
39860  39860   Williamsbridge/Olinville  Wednesday     10.0          1
39861  39861   Williamsbridge/Olinville   Thursday      9.0          1
39862  39862   Williamsbridge/Olinville     Friday      8.0          1
39863  39863   Williamsbridge/Olinville     Friday     17.0          1
39864  39864         Woodlawn/Wakefield     Monday      9.0          1
39865  39865         Woodlawn/Wakefield    Tuesday     10.0          1
39866  39866         Woodlawn/W

In [66]:
#Summarize the ridecounts by dropoff zones, and calcuate a rank by ridecount
import numpy as np

summary = pd.DataFrame(pd.pivot_table(df, values='ridecount', index=['dropoff_zone'],aggfunc=np.sum))
summary = summary.sort_values(by='ridecount',ascending=False)
summary['rank'] = np.arange(summary.shape[0])+1
summary

ridecount  rank
dropoff_zone                                                  
Midtown Center                                    456367     1
Times Sq/Theatre District                         385849     2
Murray Hill                                       381494     3
Midtown East                                      376738     4
Penn Station/Madison Sq West                      339835     5
Upper East Side South                             334615     6
Union Sq                                          332632     7
Upper East Side North                             330459     8
Clinton East                                      314599     9
East Village                                      300551    10
Midtown North                                     288713    11
Midtown South                                     268810    12
East Chelsea                                      268013    13
Lincoln Square East                               263386    14
Gramercy                                          245290    15
Lenox Hill West                                   242337    16
Upper West Side South                             237176    17
West Chelsea/Hudson Yards                         223234    18
TriBeCa/Civic Center                              204429    19
West Village                                      193484    20
Garment District                                  191947    21
Lenox Hill East                                   189348    22
Yorkville West                                    188133    23
Sutton Place/Turtle Bay North                     185979    24
Upper West Side North                             183976    25
UN/Turtle Bay South                               171774    26
Flatiron                                          166013    27
Greenwich Village North                           161851    28
Meatpacking/West Village West                     156353    29
Kips Bay                                          154943    30
...                                                  ...   ...
Heartland Village/Todt Hill                          229   231
Saint George/New Brighton                            208   232
Arrochar/Fort Wadsworth                              194   233
Stapleton                                            187   234
South Beach/Dongan Hills                             175   235
Westerleigh                                          164   236
City Island                                          163   237
Green-Wood Cemetery                                  162   238
Pelham Bay Park                                      147   239
Grymes Hill/Clifton                                  144   240
Breezy Point/Fort Tilden/Riis Beach                  140   241
West Brighton                                        139   242
Marine Park/Floyd Bennett Field                      123   243
New Dorp/Midland Beach                               114   244
Great Kills                                          112   245
Willets Point                                         99   246
Mariners Harbor                                       92   247
Crotona Park                                          81   248
Oakwood                                               79   249
Eltingville/Annadale/Prince's Bay                     70   250
Arden Heights                                         57   251
Rossville/Woodrow                                     55   252
Port Richmond                                         41   253
Charleston/Tottenville                                40   254
Broad Channel                                         39   255
Governor's Island/Ellis Island/Liberty Island          6   256
Jamaica Bay                                            5   257
Freshkills Park                                        2   258
Rikers Island                                          1   259
zone_1                                                 1   260

[260 rows x 2 columns]

In [67]:
#Save the resulting table to a CSV file. This file will be used by QGIS
summary.to_csv('zoneSummary.csv')

In [68]:
#Launch QGIS and create a new project
#Choose Layer-->Add Layer-->Add Vector Layer, and import the shapefile 'taxi_zones_wgs84.shp' created in Step 0


## Taxi Zone Map
<img src='taxiZones.png'>

The above is the raw shapefile downloaded from the NYC website. We will enhance this map in subsequent steps to make it more useful.

In [ ]:
#In QGIS, install the MMQGIS plugin. You will use this plugin to perform an attribute join
#Once the plugin is installed, choose MMQGIS-->Combine-->Attributes Join from CSV File
#Choose the zoneSummary.csv file created in the previous step
#Match the dropoff_zone field in the CSV file with the zone field in the current shapefile
#Save the output to taxi_zones_wgs84_enh.shp. This file now contains the additional attributes for ride counts and ranks
#We will use these additional fields to format the map

#The ridecount and rank fields are created as string fields by default. We need to convert them to numbers
#Open the attributes table for the shapefile and choose New Field
#Choose Create New Field and Create New Virtual field of integer type call the new fields nridecount and nrank




## Apply Colors To Map

Choose 'Categorized'

Pick 'nridecount' as the column, and 'Spectral as the color ramp'

Press the 'Classify' button, followed up 'Apply'

<img src='layerColors.png'>

## Apply Labels To Map


### Bird's Eye View of Manhattan

[Explore this map interactively](http://qgiscloud.com/vbalasu/taxi_zones_color_by_area_midtown). Here, you can click on each neighborhood to see the ride counts.

##### Blue color = More taxi dropoffs, Labels = Rank

![Bird's Eye View](BirdsEyeView.png)

Looking at the above map, we determine that Midtown Manhattan is where the action is. We will zoom in further in the next step.

![Midtown Focus](MidtownFocus.png)

## Zooming In

Based on our analysis, we have shortlisted to the top 4 neighborhoods, which all happen to be right next to each other. These zones are as follows:
- Midtown Center (#1)
- Times Sq/Theatre District (#2)
- Murray Hill (#3)
- Midtown East (#4)

In the next step, we will pull the detailed taxi trip data for these 4 neighborhoods


In [69]:
%%time
#We will import the necessary Python libraries in this step. The %%time command keeps track of the execution time for each step
import sqlite3         # Provides powerful relational database query capabilities using the SQL language
import pandas as pd    # Pandas provides a powerful DataFrame to manipulate and analyze tabular data in memory

Wall time: 1 ms


In [71]:
%%time
#We connect to a SQLite database. This database was prepared using the notebook "00 Prepare Taxi Trip Data"
#We will examine the contents of this database by looking at the sqlite_master table
cn = sqlite3.connect('../taxiJul.db')

Wall time: 29.1 ms


In [72]:
#We are interested in the taxiJulEnrich view
#Read the first row of this view to examine the columns available
sample = pd.read_sql_query("SELECT * from taxiJulEnrich LIMIT 1;",cn)
sample.columns

Index(['VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
       'passenger_count', 'trip_distance', 'pickup_longitude',
       'pickup_latitude', 'RatecodeID', 'store_and_fwd_flag',
       'dropoff_longitude', 'dropoff_latitude', 'payment_type', 'fare_amount',
       'extra', 'mta_tax', 'tip_amount', 'tolls_amount',
       'improvement_surcharge', 'total_amount', 'pickup_OBJECTID',
       'pickup_Shape_Leng', 'pickup_Shape_Area', 'pickup_zone',
       'pickup_LocationID', 'pickup_borough', 'dropoff_OBJECTID',
       'dropoff_Shape_Leng', 'dropoff_Shape_Area', 'dropoff_zone',
       'dropoff_Location', 'dropoff_borough', 'count', 'pu_year', 'pu_month',
       'pu_day', 'pu_hour', 'pu_minute', 'pu_second', 'pu_weekday', 'do_year',
       'do_month', 'do_day', 'do_hour', 'do_minute', 'do_second', 'do_weekday',
       'pu_latlong'],
      dtype='object')

In [73]:
%%time
#In this step, we use SQLite to group the taxi trips by dropoff zone, and return the counts
#Due to the large size of the dataset (11 million+ records), it is more efficient to process using SQLite rather than 
#load everything into memory
#We calculate the number of trips by dropoff zone, weekday and hour, and sort the results in descending order
#- Midtown Center (#1)
#- Times Sq/Theatre District (#2)
#- Murray Hill (#3)
#- Midtown East (#4)
df = pd.read_sql_query("SELECT * FROM taxiJulEnrich WHERE `dropoff_zone` IN ('Midtown Center','Times Sq/Theater District','Murray Hill','Midtown East');", cn)

Wall time: 53.6 s


In [74]:
df.describe().transpose()

count       mean          std        min       25%  \
passenger_count        1214599.0   1.680424     1.349408   0.000000  1.000000   
trip_distance          1214599.0   4.888817  3010.647175   0.000000  0.900000   
fare_amount            1214599.0  11.023237     8.146214 -80.000000  6.500000   
extra                  1214599.0   0.242556     0.358181  -1.000000  0.000000   
mta_tax                1214599.0   0.499369     0.021012  -0.500000  0.500000   
tip_amount             1214599.0   1.457989     2.135993 -41.000000  0.000000   
tolls_amount           1214599.0   0.244104     1.273687  -5.540000  0.000000   
improvement_surcharge  1214599.0   0.299822     0.010007  -0.300000  0.300000   
total_amount           1214599.0  13.767449    10.278073 -80.300000  8.300000   
pickup_Shape_Leng      1214599.0   0.046963     0.030995   0.024696  0.035270   
pickup_Shape_Area      1214599.0   0.000137     0.000264   0.000027  0.000056   
dropoff_Shape_Leng     1214599.0   0.038768     0.004743   0.035270  0.035270   
dropoff_Shape_Area     1214599.0   0.000065     0.000012   0.000048  0.000048   
count                  1214599.0   1.000000     0.000000   1.000000  1.000000   

                             50%        75%           max  
passenger_count         1.000000   2.000000  9.000000e+00  
trip_distance           1.400000   2.200000  3.318000e+06  
fare_amount             9.000000  12.000000  8.140000e+02  
extra                   0.000000   0.500000  1.500000e+00  
mta_tax                 0.500000   0.500000  5.000000e-01  
tip_amount              1.080000   2.060000  6.613800e+02  
tolls_amount            0.000000   0.000000  5.914800e+02  
improvement_surcharge   0.300000   0.300000  3.000000e-01  
total_amount           11.160000  14.800000  8.148000e+02  
pickup_Shape_Leng       0.041514   0.046108  2.905560e-01  
pickup_Shape_Area       0.000075   0.000109  2.195566e-03  
dropoff_Shape_Leng      0.035804   0.045769  4.576901e-02  
dropoff_Shape_Area      0.000072   0.000074  7.431507e-05  
count                   1.000000   1.000000  1.000000e+00

In [75]:
df.columns

Index(['VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
       'passenger_count', 'trip_distance', 'pickup_longitude',
       'pickup_latitude', 'RatecodeID', 'store_and_fwd_flag',
       'dropoff_longitude', 'dropoff_latitude', 'payment_type', 'fare_amount',
       'extra', 'mta_tax', 'tip_amount', 'tolls_amount',
       'improvement_surcharge', 'total_amount', 'pickup_OBJECTID',
       'pickup_Shape_Leng', 'pickup_Shape_Area', 'pickup_zone',
       'pickup_LocationID', 'pickup_borough', 'dropoff_OBJECTID',
       'dropoff_Shape_Leng', 'dropoff_Shape_Area', 'dropoff_zone',
       'dropoff_Location', 'dropoff_borough', 'count', 'pu_year', 'pu_month',
       'pu_day', 'pu_hour', 'pu_minute', 'pu_second', 'pu_weekday', 'do_year',
       'do_month', 'do_day', 'do_hour', 'do_minute', 'do_second', 'do_weekday',
       'pu_latlong'],
      dtype='object')

In [111]:
%%time
#Convert longitude and latitude to numeric, and store them as X and Y columns. Delete the text columns
latlong = df[['dropoff_longitude','dropoff_latitude']]
pd.options.mode.chained_assignment = None  # default='warn'
latlong['X'] = pd.to_numeric(df['dropoff_longitude'])
latlong['Y'] = pd.to_numeric(df['dropoff_latitude'])
del latlong['dropoff_longitude']
del latlong['dropoff_latitude']

Wall time: 1.19 s


In [112]:
latlong.describe().transpose()

count       mean       std        min        25%        50%        75%  \
X  1214599.0 -73.976404  0.003660 -73.984108 -73.979027 -73.976662 -73.973984   
Y  1214599.0  40.754370  0.005015  40.741974  40.751091  40.755047  40.758415   

         max  
X -73.966576  
Y  40.763641

In [113]:
%%time
from sklearn.cluster import KMeans
kmeans = KMeans(n_clusters=10)
model = kmeans.fit(latlong)
print("model\n", model)

model
 KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
    n_clusters=10, n_init=10, n_jobs=1, precompute_distances='auto',
    random_state=None, tol=0.0001, verbose=0)
Wall time: 1min 10s


In [114]:
model.cluster_centers_

array([[-73.97300935,  40.75616031],
       [-73.97963543,  40.75094798],
       [-73.9808618 ,  40.74500483],
       [-73.98137698,  40.75572842],
       [-73.97566408,  40.75185393],
       [-73.96974338,  40.76025142],
       [-73.9754983 ,  40.74690388],
       [-73.97432196,  40.75980209],
       [-73.97821989,  40.76021787],
       [-73.97715866,  40.75544863]])

In [115]:
%%time
predict = kmeans.predict(latlong)
latlong['group'] = pd.Series(predict, index=latlong.index)

Wall time: 271 ms


In [116]:
latlong

X          Y  group
0       -73.975922  40.757702      9
1       -73.978607  40.761799      8
2       -73.977631  40.747368      6
3       -73.979843  40.749844      1
4       -73.972984  40.755630      0
5       -73.975082  40.752052      4
6       -73.972618  40.756092      0
7       -73.972862  40.755867      0
8       -73.978050  40.745770      6
9       -73.972054  40.756809      0
10      -73.974846  40.757645      7
11      -73.973282  40.752743      4
12      -73.979065  40.752804      1
13      -73.974640  40.756844      0
14      -73.983109  40.755848      3
15      -73.972443  40.746216      6
16      -73.975395  40.752232      4
17      -73.969589  40.760967      5
18      -73.975151  40.757603      7
19      -73.982536  40.756927      3
20      -73.976830  40.752583      4
21      -73.977829  40.763401      8
22      -73.978790  40.761742      8
23      -73.967369  40.760502      5
24      -73.980263  40.750046      1
25      -73.981422  40.749828      1
26      -73.974731  40.753193      4
27      -73.977951  40.745880      6
28      -73.981255  40.752823      1
29      -73.975716  40.752071      4
...            ...        ...    ...
1214569 -73.976700  40.751839      4
1214570 -73.977722  40.758831      8
1214571 -73.979050  40.757553      8
1214572 -73.978867  40.758686      8
1214573 -73.981010  40.750237      1
1214574 -73.969391  40.758335      5
1214575 -73.983688  40.743553      2
1214576 -73.975075  40.762543      7
1214577 -73.977119  40.750954      4
1214578 -73.971924  40.760197      5
1214579 -73.977196  40.762661      8
1214580 -73.976631  40.759876      8
1214581 -73.977638  40.753647      9
1214582 -73.981689  40.752274      1
1214583 -73.974602  40.753929      4
1214584 -73.976097  40.760410      7
1214585 -73.974472  40.754932      0
1214586 -73.972740  40.746639      6
1214587 -73.969269  40.761906      5
1214588 -73.975151  40.761066      7
1214589 -73.981819  40.757366      3
1214590 -73.979439  40.755154      3
1214591 -73.981148  40.750504      1
1214592 -73.974968  40.755985      0
1214593 -73.976234  40.761322      8
1214594 -73.977119  40.758636      8
1214595 -73.981911  40.756355      3
1214596 -73.977379  40.755611      9
1214597 -73.979042  40.758713      8
1214598 -73.980186  40.748360      1

[1214599 rows x 3 columns]

In [131]:
#Create a pivot table of the cluster groups, and calculate counts of rides for each cluster
#Then order the cluster in descending order of counts
summary = latlong.pivot_table(index=['group'])
counts = latlong.pivot_table(index=['group'],aggfunc=np.size)
summary['count'] = counts['X']
summary = summary.sort_values(['count'],ascending=False)
summary

X          Y     count
group                                
1     -73.979636  40.750945  181000.0
0     -73.973013  40.756159  151770.0
4     -73.975666  40.751851  138650.0
9     -73.977164  40.755444  129451.0
5     -73.969744  40.760251  122037.0
7     -73.974322  40.759801  114694.0
8     -73.978219  40.760216  111179.0
2     -73.980861  40.745004   99016.0
3     -73.981379  40.755728   83972.0
6     -73.975494  40.746900   82830.0

# Pick the top 3 locations

Based on the matrix above, we know that the top 3 clusters are Group 1, Group 0 and Group 4

We will use the latitude and longitude of the cluster centers, not the averages shown in the above table

So the top 3 locations are as follows:
1. Group 1 : Longitude:-73.97963543,  Latitude:40.75094798
1. Group 0 : Longitude:-73.97300935, Latitude:40.75616031
1. Group 4 : Longitude:-73.97566408, Latitude:40.75185393

# And The Winners Are ...

A quick visit to Google Maps tells us that these correspond to the following street locations:
1. 40th Street, Between Madison Ave and Park Ave (40.75094798,-73.97963543)
1. 49th Street and Lexington Ave (40.75616031,-73.97300935)
1. 43rd Street and Lexington Ave (40.75185393,-73.97566408)

### #1 40th Street, Between Madison Ave and Park Ave
![40th Street, Between Madison Ave and Park Ave](Rank1.png)

### #2 49th Street and Lexington Ave
![49th Street and Lexington Ave](Rank2.png)

### #3 43rd Street and Lexington Ave
![43rd Street and Lexington Ave](Rank3.png)

We have successfully identified the top 3 locations based on our traffic and neighborhood selections. The business owner now has a solid data-driven foundation to make the location decision. Of course, there will be other factors such as rent, crime rate, etc. that should also be considered, but we now have a solid start. 